File: ios/Reguerta/Reguerta/Data/Startup/FirestoreStartupVersionPolicyRepository.swift
import FirebaseCore
import FirebaseFirestore
import Foundation

actor FirestoreStartupVersionPolicyRepository: StartupVersionPolicyRepository {
    private let storedDB: Firestore

    init(firebaseAppName: String) {
        guard let app = FirebaseApp.app(name: firebaseAppName) else {
            preconditionFailure("Firebase app is required for startup version policy")
        }
        self.storedDB = Firestore.firestore(app: app)
    }

    func policy(for platform: StartupPlatform, environment: SessionEnvironment) async throws -> StartupVersionPolicy {
        do {
            try Task.checkCancellation()
            let snapshot = try await storedDB
                .reguertaDocument(.publicConfiguration, in: .config, environment: environment)
                .getDocument(source: .server)

            guard snapshot.exists else { throw RepositoryError.notFound(resource: "config.public") }
            guard let data = snapshot.data() else {
                throw RepositoryError.invalidData(resource: "config.public.versions.\(platform.rawValue)")
            }
            return try Self.policy(for: platform, data: data)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "config.public")
        }
    }

    nonisolated static func policy(for platform: StartupPlatform, data: [String: Any]) throws -> StartupVersionPolicy {
        guard let versions = data["versions"] as? [String: Any],
              let platformPolicy = versions[platform.rawValue] as? [String: Any],
              let currentVersion = platformPolicy.requiredString(for: "current"),
              let minimumVersion = platformPolicy.requiredString(for: "min"),
              let storeURL = platformPolicy.requiredString(for: "storeUrl"),
              let forceUpdate = platformPolicy["forceUpdate"] as? Bool
        else {
            throw RepositoryError.invalidData(resource: "config.public.versions.\(platform.rawValue)")
        }

        return StartupVersionPolicy(
            currentVersion: currentVersion,
            minimumVersion: minimumVersion,
            forceUpdate: forceUpdate,
            storeURL: storeURL
        )
    }
}

nonisolated private extension Dictionary where Key == String, Value == Any {
    func requiredString(for key: String) -> String? {
        guard let value = self[key] as? String else { return nil }
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return normalized.isEmpty ? nil : normalized
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/startup/FirestoreStartupVersionPolicyRepository.kt
package com.reguerta.user.data.startup

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Source
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreDocument
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.startup.StartupPlatform
import com.reguerta.user.domain.startup.StartupVersionPolicy
import com.reguerta.user.domain.startup.StartupVersionPolicyRepository
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.tasks.await

class FirestoreStartupVersionPolicyRepository(
    private val firestore: FirebaseFirestore,
    private val environment: ReguertaFirestoreEnvironment? = null,
) : StartupVersionPolicyRepository {
    private val firestorePath = ReguertaFirestorePath(environment = environment)

    private val publicConfigDocumentPath: String
        get() = firestorePath.documentPath(
            collection = ReguertaFirestoreCollection.CONFIG,
            documentId = ReguertaFirestoreDocument.PUBLIC.wireValue,
        )

    override suspend fun getPolicy(platform: StartupPlatform): StartupVersionPolicy {
        try {
            val snapshot = firestore.document(publicConfigDocumentPath).get(Source.SERVER).await()
            if (!snapshot.exists()) {
                throw RepositoryException(
                    kind = RepositoryErrorKind.NOT_FOUND,
                    resource = STARTUP_VERSION_POLICY_RESOURCE,
                )
            }

            val versions = snapshot.get("versions") as? Map<*, *>
                ?: throw invalidStartupVersionPolicy()
            val platformPolicy = versions[platform.wireKey] as? Map<*, *>
                ?: throw invalidStartupVersionPolicy()

            val currentVersion = platformPolicy["current"].asRequiredString()
                ?: throw invalidStartupVersionPolicy()
            val minimumVersion = platformPolicy["min"].asRequiredString()
                ?: throw invalidStartupVersionPolicy()
            val storeUrl = platformPolicy["storeUrl"].asRequiredString()
                ?: throw invalidStartupVersionPolicy()
            val forceUpdate = platformPolicy["forceUpdate"] as? Boolean
                ?: throw invalidStartupVersionPolicy()

            return StartupVersionPolicy(
                currentVersion = currentVersion,
                minimumVersion = minimumVersion,
                forceUpdate = forceUpdate,
                storeUrl = storeUrl,
            )
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = STARTUP_VERSION_POLICY_RESOURCE)
        }
    }
}

private const val STARTUP_VERSION_POLICY_RESOURCE = "config.public.versions"

private fun invalidStartupVersionPolicy() = RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = STARTUP_VERSION_POLICY_RESOURCE,
)

private fun Any?.asRequiredString(): String? {
    val raw = this as? String ?: return null
    val value = raw.trim()
    return value.takeIf { it.isNotEmpty() }
}
